test/test_internet.py

```diff
@@ -9,22 +9,61 @@
 import pytest
 import warnings
 from urllib.error import URLError
+from urllib.request import Request
 
 import torchvision.datasets.utils as utils
 from common_utils import get_tmp_dir
 
 
-class TestDatasetUtils:
+@pytest.fixture
+def patch_url_redirection(mocker):
+    class Response:
+        def __init__(self, url):
+            self.url = url
+
+    def factory(*urls):
+        class PatchedOpener:
+            def __init__(self, request_or_url, *args, **kwargs):
+                self._request_or_url = request_or_url
+
+            def __enter__(self):
+                url = (
+                    self._request_or_url.full_url
+                    if isinstance(self._request_or_url.full_url, Request)
+                    else self._request_or_url.full_url
+                )
+
+                if url == urls[-1]:
+                    redirect_url = url
+                else:
+                    redirect_url = urls[urls.index(url) + 1]
```

Comment:
looks like we only ever call this with 2 urls. Could we simplify this to reflect that fact? Also above maybe we could just be `def factory(url, redirect_url):`?

Reply:
Same as https://github.com/pytorch/vision/pull/4197#discussion_r673915005.